onnxruntime/core/graph/graph_utils.cc

```diff
@@ -349,25 +309,111 @@ bool IsOutputUsed(const Node& node, int index) {
   return false;
 }
 
-bool RemoveNode(Graph& graph, Node& node) {
-  // Cannot remove a node with implicit inputs, whose output is also a graph output,
-  // or with more than one of its outputs as input to downstream Operators.
-  if (!node.ImplicitInputDefs().empty() ||
-      graph.IsNodeOutputsInGraphOutputs(node) || !IsOnlyOneOutputUsed(node)) {
+bool CanRemoveNodeAndMergeEdges(const Graph& graph, const Node& node) {
+  const std::string* output_name = nullptr;
+  if (!IsOnlyOneOutputUsed(graph, node, output_name)) {
+    return false;
+  }
+
+  // TODO: Currently we remove the node and use the input name from the node being removed.
+  // It may also be possible to instead update an upstream node to use the output name from the node being removed.
+  // This would allow removal of a node that is providing a graph output, as that output name would come from updating
+  // the upstream node. This should also enable removal if CanUpdateImplicitInputNameInSubgraphs returns false.
+
+  if (!graph.GetNodeOutputsInGraphOutputs(node).empty()) {
     return false;
   }
 
+  bool can_remove = false;
+  const std::string* new_name = nullptr;
+
   if (node.GetInputEdgesCount() == 1) {
-    // If there is a single input edge from another node (initializers are not connected with edges to nodes).
+    // we will merge the single input edge with the edges for the output that is used
+    new_name = &GetNodeInputName(node, node.InputEdgesBegin()->GetDstArgIndex());
+  } else if (node.InputDefs().size() == 1) {
+    // we can also handle a node with a single input from an initializer or graph input (no edges)
+    new_name = &node.InputDefs()[0]->Name();
+  } else {
+    // No other node removal is supported
+  }
+
+  if (new_name) {
+    // Check that changing the current output name to the new name won't break any subgraphs that consume it
+    std::vector<GraphEdge> output_edges = GetNodeOutputEdges(node);
+    can_remove = CanUpdateImplicitInputNameInSubgraphs(graph, output_edges, *new_name);
+  }
+
+  return can_remove;
+}
+
+bool RemoveNodeAndMergeEdges(Graph& graph, Node& node) {
+  // If there is a single input edge from another node (initializers are not connected with edges to nodes).
+  if (node.GetInputEdgesCount() == 1) {
+    // remove the node and wire its incoming node to its outgoing node/s
     return RemoveNodeWithSingleNodeIn(graph, node);
   }
+
+  // single input def so replace node with that
   if (node.InputDefs().size() == 1) {
-    // If a single initializer is the only input.
-    return RemoveNodeWithSingleInitializerIn(graph, node);
-  } else {
-    // No other node removal is supported, because there will be no way to connect its inputs to its outputs.
+    return ReplaceNodeWithInitializer(graph, node, *node.MutableInputDefs()[0]);
+  }
+
+  ORT_THROW("Should be unreachable if CanRemoveNodeAndMergeEdges is in sync with the logic here.");
```

Comment:
maybe this is a very implementation detailed error message? I mean mentioning CanRemoveNodeAndMergeEdges will not say much to an end user.

Reply:
This indicates an internal logic error that should never happen. It's not something a user can resolve by changing their input so the message is intended for ORT developers.